modules/tile-converter/src/i3s-converter/helpers/progress.ts

```diff
@@ -0,0 +1,130 @@
+  /**
+   * The number of digits to appear after decimal point in the string representation of the count of steps already done.
+   * It's calculated based on the total count of steps.
+   */
+  private numberOfDigitsInPercentage: number = 0;
+
+  constructor() {}
+
+  /** Total amount of work, e.g. number of files to save or number of bytes to send */
+  get stepsTotal() {
+    return this._stepsTotal;
+  }
+
+  set stepsTotal(stepsTotal) {
+    this._stepsTotal = stepsTotal;
+    this.numberOfDigitsInPercentage =
+      this.stepsTotal > 100 ? Math.ceil(Math.log10(this.stepsTotal)) - 2 : 0;
+  }
+
+  /** Amount of work already done */
+  get stepsDone() {
+    return this._stepsDone;
+  }
+
+  set stepsDone(stepsDone) {
+    this._stepsDone = stepsDone;
+  }
+
+  /**
+   * Saves the current time as we start monitoring the process.
+   */
+  startMonitoring() {
+    this.startTime = process.hrtime.bigint();
+    this.stopTime = 0n;
+  }
+
+  /**
+   * Saves the current time as we stop monitoring the process.
+   */
+  stopMonitoring() {
+    this.stopTime = process.hrtime.bigint();
+  }
+
+  /**
+   * Gets percentage of the work already done.
+   * @returns percentage of the work already done.
+   */
+  getPercent(): number | null {
+    if (!this._stepsTotal) {
+      return null;
+    }
+    const percent = (this._stepsDone / this._stepsTotal) * 100.0;
+    return percent;
+  }
+
+  /**
+   * Gets string representation of percentage of the work already done.
+   * @returns string representation of percentage
+   */
+  getPercentString() {
+    const percent = this.getPercent();
+    return percent !== null ? percent.toFixed(this.numberOfDigitsInPercentage) : '';
+  }
+
+  /**
+   * Gets the time elapsed since the monitoring started
+   * @returns Number of seconds elapsed
+   */
+  getTimeElapsed(): number {
+    const currentTime = this.stopTime ? this.stopTime : process.hrtime.bigint();
+    const diff = currentTime - this.startTime;
+    const nanoSecInSec = BigInt(1e9);
+    return Number(diff / nanoSecInSec);
+  }
+
+  /**
+   * Gets the time presumably remaining to complete the work
+   * @returns Number of seconds remaining
+   */
+  getTimeRemaining(): {timeRemaining: number; trust: boolean} | null {
+    if (!this._stepsDone || !this.startTime) {
+      return null;
+    }
+    const currentTime = this.stopTime ? this.stopTime : process.hrtime.bigint();
+    const diff = currentTime - this.startTime;
+
+    const nanoSecForOneItem = diff / BigInt(this._stepsDone);
+
+    const trust = this.isVelocityTrust(nanoSecForOneItem, this.nanoSecForOneItem);
+    this.nanoSecForOneItem = nanoSecForOneItem;
+
+    const nanoSecInSec = BigInt(1e9);
+    const timeRemainingInSeconds =
+      (BigInt(this._stepsTotal - this._stepsDone) * nanoSecForOneItem) / nanoSecInSec;
+    return {timeRemaining: Number(timeRemainingInSeconds), trust: trust};
+  }
+
+  /**
+   * Check if the computed velociy of the process can be considered trust.
```

Comment:
What does this mean? Whether we can believe the numbers we get? Maybe rephrase so it is easier to understand?

Reply:
Done